base/range.jl

```diff
@@ -53,27 +56,54 @@ function _colon(start::T, step, stop::T) where T
 end
 
 """
-    range(start, [step], length)
+    range(start; length, stop, step=1)
+
+Given a starting value, construct a range either by length or from `start` to `stop`,
+optionally with a given step (defaults to 1). One of `length` or `step` is required.
+If `length`, `stop`, and `step` are all specified, they must agree.
 
-Construct a range by length, given a starting value and optional step (defaults to 1).
+If `length` and `stop` are provided and `step` is not, the step size will be computed
+automatically such that there are `length` linearly spaced elements in the range.
 """
-range(a::T, step, len::Integer) where {T} = _range(OrderStyle(T), ArithmeticStyle(T), a, step, len)
-_range(::Ordered, ::ArithmeticWraps, a::T, step::S, len::Integer) where {T,S} =
-    StepRange{T,S}(a, step, convert(T, a+step*(len-1)))
-_range(::Any, ::Any, a::T, step::S, len::Integer) where {T,S} =
-    StepRangeLen{typeof(a+0*step),T,S}(a, step, len)
+range(start; length::Union{Integer,Nothing}=nothing, stop=nothing, step=nothing) =
+    _range(start, step, stop, length)
+
+# Bad
+_range(start, step, ::Nothing, ::Nothing) =
+    throw(ArgumentError("At least one of `length` or `stop` must be specified"))
+_range(start, ::Nothing, ::Nothing, ::Nothing) =
+    throw(ArgumentError("At least one of `length` or `stop` must be specified"))
+_range(::Nothing, ::Nothing, ::Nothing, ::Integer) =
+    throw(ArgumentError("At least one of `length` or `stop` must be specified"))
+_range(::Nothing, ::Nothing, ::Nothing, ::Nothing) =
+    throw(ArgumentError("At least one of `length` or `stop` must be specified"))
+
+# Range from start to stop
+_range(start, step,      stop, ::Nothing) = colon(start, step, stop)
+_range(start, ::Nothing, stop, ::Nothing) = colon(start, stop)
+function _range(start::Real, step::Real, stop::Real, length::Integer)
+    if !isapprox((stop - start) / (length - 1), step)
+        throw(ArgumentError("`length` and `stop` values do not agree; try passing only " *
+                            "one or the other"))
+    end
+    # We know that the passed values are consistent, so `length` is redundant
+    colon(start, step, stop)
```

Comment:
I think it'd be a little safer to discard the `step` and defer to `_range(start::T, ::Nothing, stop::T, len::Integer)`.  Steps are sloppy, lengths and endpoints can and should be exact.

Reply:
`stop` and `length` with no `step` goes through the `linspace` machinery though. Is that what we want here?

Reply:
Yes, that'd be my preference. `unique(diff(r))` likely isn't going to be `[step]` no matter what we do, but we can make the other three match exactly with a `LinRange`.

Reply:
Shouldn't this case be an error, for over-specifying?

Reply:
It is an error `if !isapprox((stop - start) / (length - 1), step)` (five lines up, just outside of the context window GitHub gives you on my comment).

Reply:
I think it should always be an error to call it this way. The `isapprox` check is going to be unpredictable, and if you get that error how do you fix your code? Probably by removing one of the arguments.